katdal/caldata.py

```diff
@@ -0,0 +1,400 @@
+    try:
+        cal_pol_ordering = np.array(cal_pol_ordering)
+    except (NameError, SyntaxError):
+        raise
+    except Exception as e:
+        raise CalibrationReadError(str(e))
+
+    if cal_pol_ordering.shape != (4, 2):
+        # assume newer telstate format
+        pol_dict = {}
+        for idx, pol in enumerate(cal_pol_ordering):
+            pol_dict[pol] = idx
+        return pol_dict
+
+    # older format needs more work
+    if cal_pol_ordering[0, 0] != cal_pol_ordering[0, 1]:
+        raise CalibrationReadError('cal_pol_ordering[0] is not consistent')
+    if cal_pol_ordering[1, 0] != cal_pol_ordering[1, 1]:
+        raise CalibrationReadError('cal_pol_ordering[1] is not consistent')
+    order = [cal_pol_ordering[0, 0], cal_pol_ordering[1, 0]]
+    if set(order) != set('vh'):
+        raise CalibrationReadError('cal_pol_ordering does not contain h and v')
+    return {order[0]: 0, order[1]: 1}
+
+
+def _get_cal_product(key, katdal_obj, **kwargs):
+    """Loads calibration solutions from a katdal file.
+
+    If an error occurs while loading the data, a warning is printed and the
+    return value is ``None``. Any keyword args are passed to
+    :func:`SimpleInterplate1D` or `ComplexInterpolate1D`.
+
+    Solutions that contain non-finite values are discarded.
+
+    Parameters
+    ----------
+    key : str
+        Name of the telescope state sensor
+
+    Returns
+    -------
+    interp : callable
+        Interpolation function which accepts timestamps and returns
+        interpolated data with shape (time, channel, pol, antenna). If the
+        solution is channel-independent, that axis will be present with
+        size 1.
+    """
+
+    timestamps, values = _get_cal_sensor(key, katdal_obj)
+
+    if values.ndim == 3:
+        # Insert a channel axis
+        values = values[:, np.newaxis, ...]
+
+    if values.ndim != 4:
+        raise CalibrationReadError('Calibration solutions has wrong number of dimensions')
+
+    # only use solutions with valid values, assuming matrix dimensions
+    # (ts, chan, pol, ant)
+    # - all values per antenna must be valid
+    # - all values per polarisation must be valid
+    # - some channels must be valid (you will interpolate over them later)
+    ts_mask = np.isfinite(values).all(axis=-1).all(axis=-1).any(axis=-1)
+    if (ts_mask.sum() / float(len(timestamps))) < 0.7:
+        raise CalibrationValueError('No finite solutions, not enough valid data')
+
+    # avoid extrapolation
+    if (timestamps[-1] < katdal_obj._timestamps[0]):
+        # All calibration solutions ahead of observation, use last one
+        values = values[[-1], ...]
+        timestamps = timestamps[[-1], ...]
+    elif (timestamps[0] > katdal_obj._timestamps[-1]):
+        # All calibration solutions after observation, use first one
+        values = values[[0], ...]
+        timestamps = timestamps[[0], ...]
+
+    values = values[ts_mask, ...]
+    timestamps = timestamps[ts_mask]
```

Comment:
I think this needs to be done before the "avoid extrapolation" section otherwise ts_mask isn't going to match timestamps/values, and the first/last one you chose might not be valid.